pkg/querier/queryrange/limits.go

```diff
@@ -5,29 +5,37 @@ import (
 	"fmt"
 	"net/http"
 	"sort"
+	"strings"
 	"sync"
 	"time"
 
+	"github.com/dustin/go-humanize"
+	"github.com/go-kit/log"
 	"github.com/go-kit/log/level"
 	"github.com/grafana/dskit/tenant"
 	"github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/model/timestamp"
 	"github.com/weaveworks/common/httpgrpc"
 	"github.com/weaveworks/common/user"
 
 	"github.com/grafana/loki/pkg/logproto"
 	"github.com/grafana/loki/pkg/logql"
+	"github.com/grafana/loki/pkg/logql/syntax"
 	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase"
 	"github.com/grafana/loki/pkg/storage/config"
+	"github.com/grafana/loki/pkg/storage/stores/index/stats"
 	"github.com/grafana/loki/pkg/util"
 	util_log "github.com/grafana/loki/pkg/util/log"
 	"github.com/grafana/loki/pkg/util/spanlogger"
 	"github.com/grafana/loki/pkg/util/validation"
 )
 
 const (
-	limitErrTmpl = "maximum of series (%d) reached for a single query"
+	limitErrTmpl                  = "maximum of series (%d) reached for a single query"
+	limErrQueryTooManyBytesTmpl   = "the query would read too many bytes (query: %s, limit: %s). Consider adding more specific stream selectors or reduce the time range of the query"
+	limErrQuerierTooManyBytesTmpl = "query too large to execute on a single querier, either because parallelization is not enabled, the query is unshardable, or a shard query is too big to execute: (query: %s, limit: %s). Consider adding more specific stream selectors or reduce the time range of the query"
```

Comment:
Can we not have different error messages when the different conditions are true? This would be quite hard for me as a querier of Loki to understand. As an operator, I wouldn't know which config option to change in order to ease this limit.

Reply:
> Can we not have different error messages when the different conditions are true?

What do you mean by "when the different conditions are true"? Do you mean when both limits are surpassed?
If so, we cannot print both errors because one is enforced before the other.

We do have different error messages for `MaxQueryBytes` and `MaxQuerierBytes`:

- `MaxQueryBytes`: `the query would read too many bytes (query: %s, limit: %s). Consider adding more specific stream selectors or reduce the time range of the query`

-  `MaxQuerierBytes`: `query too large to execute on a single querier, either because parallelization is not enabled, the query is unshardable, or a shard query is too big to execute: (query: %s, limit: %s). Consider adding more specific stream selectors or reduce the time range of the query`

Reply:
The conditions i was referring to are:

"either because parallelization is not enabled, the query is unshardable, or a shard query is too big to execute"

Reply:
That's a good point. We will implement this on a new PR.